modules/pxp-module-puppet

```diff
@@ -214,58 +229,63 @@ module Pxp
-      if last_run_report.empty?
-        return self.class.make_error_result(exitcode, Errors::NoLastRunReport,
+      if last_run_report.nil? || last_run_report.empty?
+        return make_error_result(DEFAULT_EXITCODE, Errors::NoLastRunReport,
                                  "could not find the location of the last run report")
       end
 
       # Initially ignore the lockfile. It might be out-dated, so we give Puppet a chance
       # to clean it up and run.
-      lockfile = ''
-      while true
-        wait_for_lockfile(lockfile)
-
-        start_time = get_start_time(last_run_report)
-
-        run_result = Puppet::Util::Execution.execute(puppet_agent_command, {:failonfail => false,
-                                                                            :custom_environment => env,
-                                                                            :override_locale => false})
+      start_time, exitcode = try_run(last_run_report)
+      if exitcode.nil?
+        return make_error_result(DEFAULT_EXITCODE, Errors::FailedToStart,
+                                 "Failed to start Puppet agent")
+      end
 
-        if !run_result
-          return self.class.make_error_result(exitcode, Errors::FailedToStart,
-                                   "Failed to start Puppet agent")
+      # If the run was successful, don't check for failure modes.
+      if exitcode != 0
+        if disabled?(puppet_config['agent_disabled_lockfile'] || '')
+          return make_error_result(exitcode, Errors::Disabled,
+                                   "Puppet agent is disabled")
         end
 
-        exitcode = run_result.exitstatus
-        puppet_output = force_unicode(run_result.to_s)
-
-        if disabled?(puppet_output)
-          return self.class.make_error_result(exitcode, Errors::Disabled,
-                                   "Puppet agent is disabled")
-        elsif running?(puppet_output)
-          # Puppet is already running, attempt to wait until it finishes and retry.
-          lockfile = config_print('agent_catalog_run_lockfile')
-          if lockfile.empty?
-            # Can't identify lockfile, just report an error.
-            return self.class.make_error_result(exitcode, Errors::AlreadyRunning,
-                                     "Puppet agent is already performing a run; it also appears to be" +
-                                     " misconfigured, agent_catalog_run_lockfile is an empty string")
+        # Check for a lockfile. If present, wait until it's removed and try running again.
+        # There's a chance that our run finished with a real error rather than because Puppet was
+        # already running, but another run started immediately after. Since we have no
+        # language-agnostic way to tell, we accept that we might run twice in that case.
+        # The run could also finish immediately after we tried, and the lockfile be absent.
+        # In that case we'll fail with poor error reporting.
```

Comment:
We could probably do better if we used some more sophisticated locking mechanism in puppet-agent itself.